include/dlaf/communication/communicator_grid.h

```diff
@@ -16,97 +16,134 @@
+  IndexT_MPI rankFullCommunicator(const Index2D &index) const noexcept {
+    return common::computeLinearIndex<IndexT_MPI>(
+        internal::FULL_COMMUNICATOR_ORDER, index,
+        {grid_size_.rows(), grid_size_.cols()});
+  }
+
+  std::size_t num_pipelines() const noexcept {
+    DLAF_ASSERT(full_pipelines_.size() == row_pipelines_.size(),
+                full_pipelines_.size(), row_pipelines_.size());
+    DLAF_ASSERT(full_pipelines_.size() == col_pipelines_.size(),
+                full_pipelines_.size(), col_pipelines_.size());
+    return full_pipelines_.size();
   }
 
   /// Return the rank of the current process in the CommunicatorGrid.
   ///
   /// @return the 2D coordinate representing the position in the grid
-  Index2D rank() const noexcept {
-    return position_;
-  }
+  Index2D rank() const noexcept { return position_; }
 
   /// Return the number of rows in the grid.
-  Size2D size() const noexcept {
-    return grid_size_;
-  }
+  Size2D size() const noexcept { return grid_size_; }
 
   /// Return a Communicator grouping all ranks in the grid.
-  Communicator& fullCommunicator() noexcept {
-    return full_;
-  }
+  Communicator &fullCommunicator() noexcept { return full_; }
 
-  /// Return a Communicator grouping all ranks in the row (that includes the current process).
-  Communicator& rowCommunicator() noexcept {
-    return row_;
-  }
+  /// Return a Communicator grouping all ranks in the row (that includes the
+  /// current process).
+  Communicator &rowCommunicator() noexcept { return row_; }
 
-  /// Return a Communicator grouping all ranks in the column (that includes the current process).
-  Communicator& colCommunicator() noexcept {
-    return col_;
-  }
+  /// Return a Communicator grouping all ranks in the column (that includes the
+  /// current process).
+  Communicator &colCommunicator() noexcept { return col_; }
 
-  Communicator& subCommunicator(Coord cr) noexcept {
+  Communicator &subCommunicator(Coord cr) noexcept {
     if (cr == Coord::Row)
       return row_;
     else
       return col_;
   }
 
+  /// Return a pipeline to a Communicator grouping all ranks in the grid.
+  CommunicatorPipeline<TODOCoord::Full> full_communicator_pipeline() {
+    return full_pipelines_.nextResource().sub_pipeline();
+  }
+
+  /// Return a pipeline to a Communicator grouping all ranks in the row (that
+  /// includes the current process).
+  CommunicatorPipeline<TODOCoord::Row> row_communicator_pipeline() {
+    return row_pipelines_.nextResource().sub_pipeline();
+  }
+
+  /// Return a pipeline to a Communicator grouping all ranks in the col (that
+  /// includes the current process).
+  CommunicatorPipeline<TODOCoord::Col> col_communicator_pipeline() {
+    return col_pipelines_.nextResource().sub_pipeline();
+  }
+
+  template <Coord C> auto communicator_pipeline() {
+    if constexpr (C == Coord::Row)
+      return row_communicator_pipeline();
+    else
+      return col_communicator_pipeline();
+  }
```

Comment:
`Coord` or `TODOCoord` or both as options?

Reply:
Keep only the current version since it doesn't make sense to parameterize with the option to choose `Full` (one would not need to mix `Full` with `Row`/`Col` anywhere).